### ISSUE TITLE
Add `retryAfter` to Bedrock::Jobs::GetJob

### ISSUE THREAD

**Issue by quinthar** [Expensify/Bedrock#111]
Problem:
====
Most jobs are automatically processed, such as via `BedrockWorkerManager`.  And automatically retrying a job that is automatically processed is generally a bad idea, as not knowing anything else, whatever failed the first time will likely fail again.  (And if it failed for a reason that will likely not happen again, then the worker should throw a RetryException... but that's a different story.)  

However, manually processed jobs (eg, by loading a webpage that dequeues a job for processing) often fail for reasons that are non-repeatable, such as the person who loaded the webpage to process that job just closing the browser and walking away.  In that scenario, we want to detect an "abandoned" manual job and automatically retry it.

(Note: in this scenario the website should also expire the job, such that if the web user *does* come back and try to finish it, the website should refuse it.  Otherwise we have two web users doing the same job, which ain't great.)

Solution:
====
There are likely several ways to do this, but here is my recommendation:

1. Add a `retryAfter` column to `jobs`, default 0.

2. Update `CreateJob` such that it records the new optional `retryAfter` parameter.

3. Update `GetJob` such that [the place that marks a job as `RUNNING`](https://github.com/Expensify/Bedrock/blob/master/plugins/Jobs.cpp#L363) to check if `retryAfter` is set.  If so:
    * set state=`RUNQUEUED`
        * We use a new state to signal that it is not just running, but queued to re-run in the future as well.
        * This requires updating everywhere that checks for `state="RUNNING"` to be `(state="RUNNING" || state="RUNQUEUED")`
    * set `nextRun=SCURRENT_TIMESTAMP()+retryAfter`, thereby ensuring it'll be re-run at the right time

4. Update `GetJob` (in both [`peekCommand`](https://github.com/Expensify/Bedrock/blob/master/plugins/Jobs.cpp#L84) and [`processCommand`](https://github.com/Expensify/Bedrock/blob/master/plugins/Jobs.cpp#L315)) to replace `state='QUEUED'` with `state IN ('QUEUED', 'RUNQUEUED')`

5. Update [the repeat logic in `FinishJob`](https://github.com/Expensify/Bedrock/blob/master/plugins/Jobs.cpp#L520) to set `lastScheduled = nextRun-retryAfter;`

6. To be safe, update `CreateJob` to check, if `parentJobID` is set, that the parent doesn't have a `retryAfter` set.  (If so, return `402 Auto-retrying parents cannot have children`.)  Let's not automatically retry parent jobs that create child jobs -- therein lies the path to madness.
    * Note: It's fine for an automatically processed parent to have auto-retrying manual children.  However, it's not fine for an auto-retrying manual parent to have children.

7. To be safe, update `UpdateJob` to return `402 Auto-retrying jobs cannot be updated once running` if `state='RUNQUEUED'`.   If a job is to be automatically retried, we can't have it modifying itself prior to the retry.

Performance considerations
====
I'm *guessing* this will have a negligible effect upon `GetJob` performance.  However, query plan interpretation is an art that I haven't mastered.  Here it is with the current query:

```
dbarrett@www1:~$ sudo readdb.sh "EXPLAIN QUERY PLAN
> SELECT jobID, name, data
> FROM jobs
> WHERE
>  state='QUEUED'
>  AND current_timestamp >= nextRun
>  AND name GLOB 'www-prod/*'
> ORDER BY priority DESC, nextRun ASC
> LIMIT 1;
> "
selectid    order       from        detail                                                                    
----------  ----------  ----------  --------------------------------------------------------------------------
0           0           0           SEARCH TABLE jobs USING INDEX jobsStateNextRunName (state=? AND nextRun<?)
0           0           0           USE TEMP B-TREE FOR ORDER BY                                              
```

And here it is with the proposed update:

```
dbarrett@www1:~$ sudo readdb.sh "EXPLAIN QUERY PLAN
> SELECT jobID, name, data
> FROM jobs
> WHERE
>  state IN ('QUEUED', 'RUNQUEUED')
>  AND current_timestamp >= nextRun
>  AND name GLOB 'www-prod/*'
> ORDER BY priority DESC, nextRun ASC
> LIMIT 1;
> "
selectid    order       from        detail                                                                    
----------  ----------  ----------  --------------------------------------------------------------------------
0           0           0           SEARCH TABLE jobs USING INDEX jobsStateNextRunName (state=? AND nextRun<?)
0           0           0           EXECUTE LIST SUBQUERY 1                                                   
0           0           0           USE TEMP B-TREE FOR ORDER BY                                              
dbarrett@www1:~$ 
```

Thoughts or concerns on this before I get started?  @cead22 @iwiznia @flodnv @tylerkaraszewski @mcnamamj

**Comment by quinthar** [comment]
@danielk1977 - As the resident query plan master, can you comment on whether you think the above query plan would be significantly slower than the original?  Thanks!

**Comment by danielk1977** [comment]
They're quite similar. The first scans a subset of the index jobsStateNextRunName and does a by-rowid lookup on the main table for each index row scanned. It then tests the GLOB condition, and if it matches inserts the row into the temporary b-tree. Once it has accumulated all matching rows in the temp b-tree, it reads them out in sorted order to return to the user.

The second query is similar, except that it scans two regions of the index to collect rows.

I think the relative speed of these two queries will be roughly proportional to the number of rows returned. In the worst case (no rows returned), I guess the second might take up to twice as long as the first.

Dan.

**Comment by quinthar** [comment]
Well, that query is already brutally slow so that's not great.  But it's
already "too slow" so "2x too slow" isn't really much of a change.
Regardless, we need to find a way to speed up "GetJob" overall to make it
"fast enough", and then hopefully "2x fast enough" is still fast enough.
Thanks!

**Comment by quinthar** [comment]
Actually, strike that -- it's actually super fast, it looks like a flaw in our monitoring system was mis-attributing a slowdown to Bedrock.

**Comment by fnwbr** [comment]
So what's the status with this?
The PR was reverted with https://github.com/Expensify/Bedrock/pull/194 - correct?

**Comment by fnwbr** [comment]
Bump, any news on this?
If the code itself was fine and we only had to revert because of the schema changes taking too long when done via `upgradeDatabase` can we please look into manually running the schema changes first?

**Comment by tylerkaraszewski** [comment]
We split the change into a schema change and a code change so that we could test them individually, and we're waiting for the schema change to be deployed. The fires this week have kept #infra pretty busy so it's been a bit delayed. When the schema change goes out, we'll see if that causes a problem, indicating a code problem, or if the schema change causes a problem itself.

**Comment by tylerkaraszewski** [comment]
@coleaeason - we're still waiting on the schema change to be deployed, correct?

**Comment by flodnv** [comment]
I can help with the deploys, I'm down to deploy bedrock this week.

**Comment by coleaeason** [comment]
Sorry, yeah i've had no time for this and it's pretty low priority for me compared to other things on my plate. If someone else ( @flodnv or @iwiznia ) wants to take ownership of deploying this since they are more invested in getting it out, then I'm fine with that.

**Comment by fnwbr** [comment]
Bumping this again; I know you are having lots of other priorities, this would definitely help us on the Chatbot front though.

**Comment by flodnv** [comment]
@tylerkaraszewski where's the PR for this? Is is this https://github.com/Expensify/Bedrock/pull/235 ? Then bump @cead22 @mea36

**Comment by cead22** [comment]
@flodnv yes.

@fnwbr @mea36 feel free to take over recreating the PR. It's filled with conflicts that since the PRs CancelJob and CreateJobs got merged.

Please make sure to run the tests in #235 #191 and #195 when submitting the PR
cc @pecanoro

**Comment by mea36** [comment]
Thanks for the update @cead22 
Why aren't those tests automated?

**Comment by cead22** [comment]
We hadn't started writing tests at the time and I didn't know how to setup the environment to run them. I asked @flodnv to get that started and now we have tests in place.

As a reminder, I wrote the original PR in May

**Comment by fnwbr** [comment]
Re-created the original PR (linked above) to get this kickstarted; any reason that @tylerkaraszewski's un-revert did not have the schema changes coming with the PR anymore?

I also left them out of my PR for now; also didn't test anything yet

**Comment by cead22** [comment]
The original PR included 2 changes: retryAfter and the schema change and when we deployed them we saw issues (commands being queued but not dequeued and processed) and we decided to split the two changes and deploy them separately.

The schema change is already deployed and now we want to deploy retryAfter to see if it causes issues

```
carlos@www1:~$ sudo readdb.sh ".schema jobs"
CREATE TABLE jobs ( created TIMESTAMP NOT NULL, jobID INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT, state TEXT NOT NULL, name TEXT NOT NULL, nextRun TIMESTAMP NOT NULL, lastRun TIMESTAMP, repeat TEXT NOT NULL, data TEXT NOT NULL , priority INTEGER NOT NULL DEFAULT 500, parentJobID INTEGER NOT NULL DEFAULT 0 , retryAfter TEXT NOT NULL DEFAULT "");
```

**Comment by fnwbr** [comment]
Ah, I didn't see that `retryAfter` is still in the schema for when creating a completely new `jobs` table / I didn't notice that it wasn't removed from there; so we're good 👍

**Comment by mea36** [comment]
Can someone clarify this point
> To be safe, update UpdateJob to return 402 Auto-retrying jobs cannot be updated once running if state='RUNQUEUED'. If a job is to be automatically retried, we can't have it modifying itself prior to the retry.

Why can we update the data for a `RUNNING` job but not a `RUNQUEUED` job?


cc @pecanoro

**Comment by quinthar** [comment]
I can't remember why I wrote that.  I can't think of a good explanation for that, so lacking a reason I suggest we ignore that comment.

**Comment by iwiznia** [comment]
Reopening as this had a performance regression and had to be rolled back.
